formula/src/main/java/com/instacart/formula/batch/BatchManager.kt

```diff
@@ -0,0 +1,64 @@
+package com.instacart.formula.batch
+
+import com.instacart.formula.FormulaRuntime
+import com.instacart.formula.Transition
+import com.instacart.formula.internal.SynchronizedUpdateQueue
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * The BatchManager is initialized by [FormulaRuntime] and scoped to [FormulaRuntime] root
+ * to manage batched events that happen in that formula tree. When an event marked
+ * with [Transition.Batched] arrives:
+ * - If we are already processing, we will process the event immediately without batching
+ * - Otherwise, we will check if there already exists a batch matching [BatchScheduler.key]. If one
+ * exists, we will add the event to that batch.
+ * - If no batch exists, we will initialize the batch and call [BatchScheduler.schedule] to
+ * schedule the execution of the batch. The [BatchScheduler] will keep a reference to the batch
+ * until it is executed.
+ */
+internal class BatchManager(
+    private val runtime: FormulaRuntime<*, *>,
+    private val updateQueue: SynchronizedUpdateQueue,
+) {
+    private var activeBatches: MutableMap<Any, BatchImpl>? = null
+
+
+    fun add(batched: Transition.Batched, event: Any?, update: () -> Unit) {
+        if (updateQueue.isProcessingOnCurrentThread()) {
+            /**
+             * Behavior: If we are already processing formula on this thread, let's not batch this event and
+             * process it immediately. Maybe in the future, we will want to make this behavior
+             * configurable, but for now let's set this as default behavior.
+             *
+             * Dispatching: since we are already processing on this thread, no need to dispatch
+             * it to another thread.
+             */
+            update()
+        } else {
+            val batch = synchronized(this) {
+                initOrGetBatch(batched.scheduler, event).apply {
+                    // TODO: implement drop previous event!
+                    add(update)
+                }
+            }
+            batch.scheduleIfNeeded(batched.scheduler)
+        }
+    }
+
+    fun removeBatch(batch: BatchImpl): Boolean {
+        val batchImpl = synchronized(this) {
+            activeBatches?.remove(batch.key())
+        }
+        return batchImpl === batch
+    }
+
+    private fun initOrGetBatch(scheduler: BatchScheduler, event: Any?): BatchImpl {
+        val activeBatches = activeBatches ?: mutableMapOf()
+        this.activeBatches = activeBatches
```

Comment:
Why is activeBatches nullable? Could it not be initialized to an empty map?

Reply:
It's an optimization to avoid initializing a map unless needed (many formulas do not need batching). The problem with `emptyMap` is that it's not mutable, and using `mutableMapOf` will pre-allocate memory.